component/xload/xload.go

```diff
@@ -0,0 +1,300 @@
+		log.Err("Xload::Configure : config error [invalid config attributes]")
+		return fmt.Errorf("Xload: config error [invalid config attributes]")
+	}
+
+	xl.blockSize = uint64(defaultBlockSize) * _1MB // 16 MB as deafult block size
+	if config.IsSet(compName + ".block-size-mb") {
+		xl.blockSize = uint64(conf.BlockSize * float64(_1MB))
+	}
+
+	xl.path = common.ExpandPath(strings.TrimSpace(conf.Path))
+	if xl.path == "" {
+		// TODO:: xload : should we use current working directory in this case
+		log.Err("Xload::Configure : config error [path not given in xload]")
+		return fmt.Errorf("config error in %s [path not given]", xl.Name())
+	} else {
+		//check mnt path is not same as xload path
+		mntPath := ""
+		err = config.UnmarshalKey("mount-path", &mntPath)
+		if err != nil {
+			log.Err("Xload::Configure : config error [unable to obtain Mount Path [%s]]", err.Error())
+			return fmt.Errorf("config error in %s [%s]", xl.Name(), err.Error())
+		}
+
+		if xl.path == mntPath {
+			log.Err("Xload::Configure : config error [xload path is same as mount path]")
+			return fmt.Errorf("config error in %s error [xload path is same as mount path]", xl.Name())
+		}
+
+		_, err = os.Stat(xl.path)
+		if os.IsNotExist(err) {
+			log.Info("Xload::Configure : config error [xload path does not exist, attempting to create path]")
+			err := os.Mkdir(xl.path, os.FileMode(0755))
+			if err != nil {
+				log.Err("Xload::Configure : config error creating directory of xload path [%s]", err.Error())
+				return fmt.Errorf("config error in %s [%s]", xl.Name(), err.Error())
+			}
+		}
+
+		if !common.IsDirectoryEmpty(xl.path) {
+			log.Err("Xload::Configure : config error %s directory is not empty", xl.path)
+			return fmt.Errorf("config error in %s [temp directory not empty]", xl.Name())
+		}
+	}
+
+	var mode xcommon.Mode
+	err = mode.Parse(conf.Mode)
+	if err != nil {
+		log.Err("Xload::Configure : Failed to parse mode %s [%s]", conf.Mode, err.Error())
+		return fmt.Errorf("invalid mode in xload : %s", conf.Mode)
+	}
+
+	if mode == xcommon.EMode.INVALID_MODE() {
+		log.Err("Xload::Configure : Invalid mode : %s", conf.Mode)
+		return fmt.Errorf("invalid mode in xload : %s", conf.Mode)
+	}
+
+	xl.mode = mode
+
+	return nil
+}
+
+// Start : Pipeline calls this method to start the component functionality
+func (xl *Xload) Start(ctx context.Context) error {
+	log.Trace("Xload::Start : Starting component %s", xl.Name())
+
+	xl.workerCount = xcommon.MAX_WORKER_COUNT
+	xl.blockPool = xcommon.NewBlockPool(xl.blockSize, xl.workerCount*3)
+	if xl.blockPool == nil {
+		log.Err("Xload::Start : Failed to create block pool")
+		return fmt.Errorf("failed to create block pool")
+	}
+
+	var err error
+
+	// Xload : start code goes here
+	switch xl.mode {
+	case xcommon.EMode.CHECKPOINT():
+		// Start checkpoint thread here
+		return fmt.Errorf("checkpoint is currently unsupported")
+	case xcommon.EMode.DOWNLOAD():
+		// Start downloader here
+		err = xl.startDownloader()
```

Comment:
We are doing "startDownloader" here and then at the end of the function we call "startComponents". As the later is actually starting the download part, this method shall be called something else as its preparing the pipeline for download and not actually tirggering any download at this point.

Reply:
Renamed this method to `createDownloader`.